text/0000-deletable-immutable-data/0000-deletable-immutable-data.md

```diff
@@ -0,0 +1,68 @@
+struct ImmutableData(Vec<u8>);
+let immut_data = ImmutableData::new(vec![255; 10]);
+
+// Name:
+let immut_data_name = XorName::new(sha512::hash(&immut_data.0).0);
+
+// Location on SAFE Network: stored by group of nodes close to immut_data_name
+
+//Conceptual storage in a node:
+let storage: HashMap<XorName, ImmutableData>;
+```
+For someone to be able to delete it there has to be some notion of ownership tied to it, otherwise `B` can delete data that was created only by `A`. We need a mechanism in place where `B` can delete data created by `A` iff `B` too had created exactly the same data. After `B` has done deleting, the data should still remain as `A` hasn't deleted it yet. Incentivising would mean that whoever deletes data gets reimbursed for playing their part in freeing up space. From this point of view, we don't want `B's` subsequent delete attempts to be legit (allowed), because `B` has already been reimbursed. When `A` finally issues a delete, a reimbursement is credited and data is finally deleted by the network.
+
+Changing what is mentioned as a conceptual storage of `ImmutableData` in a node above can help achive this:
+```rust
+enum Owner {
+    Users(Vec<sign::PublicKey>),
+    Network,
+}
+
+struct ImmutableDataStorage {
+    immut_data: ImmutableData,
+    owner: Owner,
+}
+
+// New way of storage:
+let new_storage: HashMap<XorName, ImmutableDataStorage>;
+```
+Each time someone creates the same `ImmutableData`, one's `PublicKey` get appended to the `Owner::Users` field. Similarly each delete request from a user will be considered valid if the user's key is present in the list and if so the found key will be removed from the list. Once the length of `Owner::Users` reduces to zero, the data is deleted from the network. This field must allow duplication because one user can create same `ImmutableData` multiple times. For e.g. if `A` creates same `ImmutableData` thrice, `A` can issue 3 deletes all of which will be successful and will yeild a reimbursement. The last delete will actually delete the data from the network.
+
+There is however a downside - for popular data which might be deduplicated a large number of times, `Owner::Users` will grow enormously in size. A solution to this is that deletion of `ImmutableData` must be entail a _best-effort_ guarantee. If the deduplication rises above a limit (hence highly popular), we will consider that data to be entirely network owned and can no longer be deleted by any of its creators. At this point the storage will flip `owner` to `Owner::Network` and beyond this, `ImmutableData` can no longer be deleted and has no associated owners. The policy of reimbursement will not be applicable anymore.
+
+This upper limit shall be called `DEDUPLICATION_LIMIT` and shall be defined as follows: _The number of unique keys in `Owner::Users` beyond which this `ImmutableData` is entirely owned by the network_.
+Its value is chosen thus:
+```rust
+const DEDUPLICATION_LIMIT: usize = 20;
```

Comment:
Another nice feature of this is that a potential vandal trying to fill the network with rubbish that does not delete would need to pay for it many time over (20 here). 

Of course it does not answer the other question about a vandal uploading stuff that does not get deleted (i.e. they never call delete anyway, since they are attempting vandalism)